scripts/cos_similarity_correlation_test.py

```diff
@@ -0,0 +1,149 @@
+the user reviews map and performs randomization on the data. A chi square value is
+calculated on the data set as a measure of correlation. This value is calculated
+for each individual restaurant when possible and summed together to form a larger
+chi square statistic.
+
+Restaurants where expected value in the chisq contingency table is 0 are ignored.
+
+This is repeated RAND_NUM times on randomized data to create a distribution,
+and then the percentile of the actual chi square statistic for the real data set
+is determined based on the randomized data to determine significance.
+
+Argument 1: RAND_NUM
+
+Argument 2: 'score' or 'edge' to determine if reviews should be randomized or
+    edges in the friend graph
+'''
+
+from correlation_test import (remove_dups_to_dict, get_user_combos_from_ratings_map,
+    get_random_score_ratings, get_random_edge_graphs, get_edge_set)
+from init_friend_graph import init_friend_graph
+from itertools import combinations
+import pickle
+import random
+from random import shuffle
+import sys
+
+RAND_NUM = int(sys.argv[1])
+
+
+def main():
+    rest_user_reviews_map = read_restaurant_by_user_reviews()
+    num_restaurants = len(rest_user_reviews_map)
+    friend_graph = init_friend_graph('../data/yelpFriends.txt')
+
+    # A list of lists, where each list corresponds to a restaurant subgraph
+    # Each list contains RAND_NUM chi sq values generated from permuting the graph
+    random_chisq_values = []
+
+    # Running sum of the chi sq values for each restaurant of the real graph
+    real_chisq_sum = 0
+
+    counter = 0
+
+    for r_id in rest_user_reviews_map:
+        user_reviews = rest_user_reviews_map[r_id]
+        combinations = get_user_combos_from_ratings_map(user_reviews['users'])
+        # generate random chisq values first:
+        chisq_vals = gen_random_chisq_vals(user_reviews, friend_graph, combinations)
+        #print("calculated " + str(counter) + " of " + str(num_restaurants))
+        counter += 1
+        random_chisq_values.append(chisq_vals)
+
+        # calculate chisq value for actual graph
+        real_chisq_sum += calc_chi_sq(user_reviews['users'], user_reviews['cos_matrix'], friend_graph, combinations)
+
+    # list of chisq sums for RAND_NUM trials
+    sum_rand_chisq = [sum(i) for i in zip(*random_chisq_values)]
+    sum_rand_chisq.sort()
+
+    for num in sum_rand_chisq:
+        print(str(num))
+    print(real_chisq_sum)
+    # calculate what percentile the real sum is in compared to the distribution of random
+    percentile = 1
+    for i in range(len(sum_rand_chisq)):
+        if sum_rand_chisq[i] > real_chisq_sum:
+            percentile = i/len(sum_rand_chisq)
+            break
+    print("Percentile: " + str(percentile))
+
+
+def read_restaurant_by_user_reviews():
+    '''
+    Reads user ratings grouped by restaurants
+    '''
+    with open('../data/cosSimilarityNoNullRestaurants.pkl', 'rb') as restaurant_f:
+        user_reviews_map = pickle.load(restaurant_f)
```

Comment:
Should we close the file at some point after we load the pickle or does with do that automatically?

Reply:
Yeah “with” closes the file automatically